### DISCUSSION TITLE
Does the firmware run on a Heltec WiFi Kit 32?

### DISCUSSION THREAD

**Discussion by fkroepfl** [UtilitechAS/amsreader-firmware#801] [General]
I would like to use the firmware on a Heltec WiFi Kit 32 [ESP Chip ID | 15053036 (ESP32-S3 v0.2)].

Is this board supported and if so, which FW variant is the right one?

Many thanks in advance.

**Comment by ArnieO** [comment]
Yes, that should work, but:

1. The OLED will not show anything
2. You must use an available GPIO (not used by the board already) for the incoming signal from the meter
3. You must use binary files built for ESP32-S3:
![image](https://github.com/UtilitechAS/amsreader-firmware/assets/10295178/f06c17b5-b880-4bdc-a4cd-60a33b61e5b1)

**Reply by fkroepfl** [reply]
Thank you for the clarification, although I was not successful. 

I used the zip file and flashed it with https://esp.huhn.me/ and the information in the wiki for ESP32 and ESP32-S2.

So far so good, but the board refused with a kind of "boot error".

I then used a generic ESP32 board. The corresponding zip file, same flasher. That then worked.

**Reply by ArnieO** [reply]
When you say "generic ESP32 board": It is a board with ESP32-S3, right?

**Reply by fkroepfl** [reply]
Sorry, no I meant something like this 🤗

- https://www.berrybase.de/esp32-nodemcu-development-board

**Reply by ArnieO** [reply]
OK, and when you wrote "corresponding zip file" it was this ZIP file you used on that board:  ams2mqtt-esp32-2.3.5.zip 
- right?
(not  ams2mqtt-esp32s3-2.3.5.zip )

I'm asking because I'd like to know whether there could be an issue with our build for S3.

**Reply by fkroepfl** [reply]
yes, I used "ams2mqtt-esp32-2.3.5.zip" for the esp32-nodemcu-development-board 🤗

The "s3" variant, I was using for the Heltec Wifi Kit

**Reply by ArnieO** [reply]
Thank you for clarifying!

Did you try an "erase flash" and reflash on the Heltec board? 
https://github.com/UtilitechAS/amsreader-firmware/wiki/flashing#erase-flash

**Reply by fkroepfl** [reply]
Not explicitly, but I always select the „Erase" option when flashing.

**Reply by ArnieO** [reply]
Could you try a separate "Erase flash", then reflashing, before we look into whether there is an issue with the build/binaries for S3?
I don't have an S3 available for test, but @gskjold surely has.

**Reply by fkroepfl** [reply]
But sure, I just have to finish testing the app that I have installed on the Heltec kit in the meantime. I hope that's OK for you

**Reply by fkroepfl** [reply]
I'm sorry, but none of my attempts to get the Heltec WiFi Kit 32 V3 to work with your FW were successful.

![image](https://github.com/UtilitechAS/amsreader-firmware/assets/6333880/e70099a0-c221-4aeb-9ac2-e31ea3072a1d)

```
Click Connect to start
Connecting...
Connected successfully.
Try hard reset.
Chip type ESP32-S3
Connected to ESP32-S3
MAC Address: 48:27:E2:E5:B0:EC
Uploading stub...
Running stub...
Stub is now running...
Detecting Flash Size
FlashId: 0x1740C8
Flash Manufacturer: c8
Flash Device: 4017
Auto-detected Flash size: 8MB
Image header, Magic=0xE9, FlashMode=0x02, FlashSizeFreq=0x3F
Writing data with filesize: 13792
Erase size 13792, blocks 1, block size 0x4000, offset 0x1000, encrypted no
Flashing... 100%
Took 1490ms to write 13792 bytes
Erase size 0, blocks 0, block size 0x4000, offset 0x0000, encrypted no
Image header, Magic=0xAA, FlashMode=0x01, FlashSizeFreq=0x02
Writing data with filesize: 3072
Erase size 3072, blocks 1, block size 0x4000, offset 0x8000, encrypted no
Flashing... 100%
Took 1447ms to write 3072 bytes
Erase size 0, blocks 0, block size 0x4000, offset 0x0000, encrypted no
Image header, Magic=0x01, FlashMode=0x00, FlashSizeFreq=0x00
Writing data with filesize: 8192
Erase size 8192, blocks 1, block size 0x4000, offset 0xE000, encrypted no
Flashing... 100%
Took 1454ms to write 8192 bytes
Erase size 0, blocks 0, block size 0x4000, offset 0x0000, encrypted no
Image header, Magic=0xE9, FlashMode=0x02, FlashSizeFreq=0x3F
Writing data with filesize: 1218144
Erase size 1218144, blocks 75, block size 0x4000, offset 0x10000, encrypted no
Flashing... 1%
...
Flashing... 100%
Took 109815ms to write 1218144 bytes
Erase size 0, blocks 0, block size 0x4000, offset 0x0000, encrypted no
Done!
To run the new firmware please reset your device.
```

**Reply by ArnieO** [reply]
OK, then there could be an issue with the binaries for S3. We had an issue also with C3 recently, maybe same problem here. We only use S2 on our products, but want to make binaries available also for other versions (with a bit lower priority). 

Please be patient until @gskjold gets a chance to look into this.

**Reply by fkroepfl** [reply]
But sure, no problem, I have an alternative ESP32 board to test if the amsreader firmware understands the data packets of the L&G E450 (Mbus) of STW-Energie Klagenfurt out of the box. 

Apparently they use a not quite usual, but still valid structuring of the meter data. Anyway, I'm already curious.

**Reply by ArnieO** [reply]
> L&G E450 (Mbus) of STW-Energie Klagenfurt

Can you point me to where I find the documentation on https://www.stw.at ?
(I can read technical Deutsch, but I struggle finding what I'm looking for.)

This map shows the postal codes of our customers, and we have quite a number in Austria - but not yet in your region I think. https://www.amsleser.no/module/ets_blog/blog?id_post=11

**Reply by Noschvie** [reply]
Please find the link below, but there is no technical description available...

https://www.energieklagenfurt.at/stromnetz/smart-meter/#landis-drehstrom

@fkroepfl maybe you send an email to smartmeter@stw.at

**Reply by ArnieO** [reply]
What we would need is technical documentation that describes the payload. Our firmware can "eat" without modification most payloads that send the DLMS/COSEM code and then the parameter - but there are some that prefer not to do it like that...

**Reply by fkroepfl** [reply]
@ArnieO @Noschvie Guys, thank you very much for your advice, tips and support in using this smart meter reader to get near real-time data from one of our smart meters.

Now I just need a level converter for the P1 interface of the smart meters of KNG-Kärnten Netz GmbH (address: Arnulfplatz 2, 9020 Klagenfurt am Wörthersee, Austria, https://kaerntennetz.at/smartmeter.htm) to see if the AMS Reader also works with the ISKRA AM550 without any problems.

It looks like there aren't too many AMS readers in operation in the south of Austria yet. But if word gets around that these modules work straight away, that will probably change soon.

<img width="1065" alt="Bildschirmfoto 2024-06-02 um 09 02 08" src="https://github.com/UtilitechAS/amsreader-firmware/assets/6333880/a8f3d276-337a-49fc-af02-275e4b43ac3b">

**Reply by fkroepfl** [reply]
> Can you point me to where I find the documentation on https://www.stw.at ?

I attached the doc I received here: https://github.com/UtilitechAS/amsreader-firmware/discussions/802#discussion-6761535

**Comment by fkroepfl** [comment]
Guys, you can find the result of the journey so far with an L&G E450 from Energie Klagenfurt (Austria) here: https://github.com/UtilitechAS/amsreader-firmware/discussions/802#discussion-6761535
